### PR TITLE
DEVPROD-6923: allow skipping known issues

### PR DESCRIPTION
**Pull request by rongodb**
Screenshot 2024-07-25 at 2 27 35 PM

See in image, the version 6501c4f2d1fe071dbddc55af has a failing task that is a known issue.
without the override, no version is found due to the failure. with the override - the version is considered passing

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
 # Changelog
+## 7.0.3 - 2024-07-25
```
Comment (zituo-jin) [inline CHANGELOG.md]: I think this should be 7.1.0 according to [semver](https://semver.org/)?

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "git-co-evg-base"
-version = "7.0.2"
+version = "7.0.3"
```
Comment (zituo-jin) [inline pyproject.toml]: same as above

File: src/goodbase/goodbase_cli.py [hunk 3]
```diff
@@ -138,6 +138,7 @@ def checkout_good_base(
         self,
         evg_project: str,
         build_checks: List[BuildChecks],
+        allow_known_failures: bool,
```
Comment (zituo-jin) [inline src/goodbase/goodbase_cli.py]: Do we need a default value here?

File: src/goodbase/goodbase_cli.py [hunk 4]
```diff
@@ -291,6 +294,7 @@ def configure_logging(verbose: bool) -> None:
     help="Number of commits to check before giving up.",
 )
 @click.option("--version-override", help="A specific version to test.")
+@click.option("--allow-known-failures", type=bool, help="consider known failures as passing.")
```
Comment (zituo-jin) [inline src/goodbase/goodbase_cli.py]: I'm thinking maybe we should use `@click.option("--allow-known-failures", is_flag=True, default=False, help="consider known failures as passing.")` instead because this is more like a flag to users right? So they can just allow known failures by passing in `--allow-known-failures` instead of `--allow-known-failures=True` (and disallow it by not specifying this option at all). WDYT?

File: src/goodbase/goodbase_cli.py [hunk 5]
```diff
@@ -361,6 +365,7 @@ def main(
     override: bool,
     verbose: bool,
     version_override: Optional[str],
+    allow_known_failures: Optional[bool] = False,
```
Comment (zituo-jin) [inline src/goodbase/goodbase_cli.py]: With the suggested change above, I think this can just be `allow_known_failures: bool`.

File: src/goodbase/services/evg_service.py [hunk 6]
```diff
@@ -32,8 +32,9 @@ def __init__(self, evg_api: EvergreenApi, evg_cli_proxy: EvgCliProxy) -> None:
         self.evg_api = evg_api
         self.evg_cli_proxy = evg_cli_proxy
 
-    @staticmethod
-    def analyze_build(build: Build) -> Optional[BuildStatus]:
+    def analyze_build(
+        self, build: Build, allow_known_failures: bool = False
```
Comment (zituo-jin) [inline src/goodbase/services/evg_service.py:36]: Don't forget to update the dostring here.

File: src/goodbase/services/evg_service.py [hunk 7]
```diff
@@ -63,15 +64,42 @@ def analyze_build(build: Build) -> Optional[BuildStatus]:
             all_tasks=all_tasks,
         )
 
-    def check_version(self, evg_version: Version, build_checks: List[BuildChecks]) -> bool:
+    def _get_successful_tasks(self, tasks: List[Task], allow_known_failures: bool) -> set[str]:
```
Comment (zituo-jin) [inline src/goodbase/services/evg_service.py:68]: [nit] I know this is not a public method, but having a docstring would help us in the future and you have added it to the private method below anyway :).

File: src/goodbase/services/evg_service.py [hunk 8]
```diff
@@ -63,15 +64,42 @@ def analyze_build(build: Build) -> Optional[BuildStatus]:
...
+        return {
+            task.display_name
+            for task in tasks
+            if task.is_success() or (allow_known_failures and self._is_task_a_known_failure(task))
+        }
+
+    def _is_task_a_known_failure(self, task: Task) -> bool:
+        """
+        Check if a task is a known failure - annotated with an issue or suspected issue that has a BF prefix.
```
Comment (zituo-jin) [inline src/goodbase/services/evg_service.py]: Looks like you've updated the implementation but not docstring.

File: src/goodbase/services/evg_service.py [hunk 9]
```diff
@@ -63,15 +64,42 @@ def analyze_build(build: Build) -> Optional[BuildStatus]:
...
+        for annotation in annotations:
+            if annotation.issues:
+                return True
+        return False
+
+    def check_version(
+        self,
+        evg_version: Version,
+        build_checks: List[BuildChecks],
+        allow_known_failures: bool = False,
```
Comment (zituo-jin) [inline src/goodbase/services/evg_service.py:103]: Don't forget the docstring for the new parameter :)

Comment (zituo-jin) [reply src/goodbase/services/evg_service.py:103]: [nit] I wonder if passing in `False` explicitly is still needed for some of the tests you've fixed below, since it already defaults to `False`?

**Review by zituo-jin — COMMENTED** [review]
Looks good overall, but I'm a little surprised `mypy`/`flake8` is not failing on docstring mismatch. Can we make sure it's updated wherever the new parameter is added?

File: src/goodbase/services/evg_service.py [hunk 10]
```diff
@@ -63,15 +64,42 @@ def analyze_build(build: Build) -> Optional[BuildStatus]:
...
+        Check if a task is a known failure - annotated with an issue or suspected issue that has a BF prefix.
+
+        :param task: the task to check.
+        :return: True if the task is a known failure.
+        """
+        annotations = self.evg_api.get_task_annotation(task.task_id)
+        for annotation in annotations:
+            if annotation.issues:
+                return True
+        return False
```
Comment (gmishkin) [inline src/goodbase/services/evg_service.py]: nit: could probably be `any(annotation.issues for annotation in annotations)` or something like that.

File: src/goodbase/services/evg_service.py [hunk 11]
```diff
@@ -63,15 +64,42 @@ def analyze_build(build: Build) -> Optional[BuildStatus]:
             all_tasks=all_tasks,
         )
 
-    def check_version(self, evg_version: Version, build_checks: List[BuildChecks]) -> bool:
+    def _get_successful_tasks(self, tasks: List[Task], allow_known_failures: bool) -> set[str]:
+        return {
+            task.display_name
+            for task in tasks
+            if task.is_success() or (allow_known_failures and self._is_task_a_known_failure(task))
```
Comment (gmishkin) [inline src/goodbase/services/evg_service.py]: I would make _really_ sure to read the Python language documentation for boolean expressions to make sure that it doesn't unconditionally call `_is_task_a_known_failure()` if allow_known_failures is false or the task succeeded 😉

**Review by gmishkin — COMMENTED** [review]
What's the performance like? I see it's already using an executor for analyzing multiple versions, does having to pull all the annotations sequentially make it a lot slower?

File: src/goodbase/services/evg_service.py (hunk 11 above)
Comment (rongodb) [reply src/goodbase/services/evg_service.py]: hmmm.....

In most languages I know they sequentially check... If the first is true, they don't continue - but I'll check.

seemingly, in python it's the same:
print("A ") if True and print("B ") else print("C ")
B
C

print("A ") if False and print("B ") else print("C ")
C

(B was not printed)

and same with or statement:
if True or print("B ") else print("C ")
A

**Comment by rongodb** [comment]
> What's the performance like? I see it's already using an executor for analyzing multiple versions, does having to pull all the annotations sequentially make it a lot slower?

TL; DR: If you're finding a version much faster, then it's surely worth the wait. If not, it's marginally slower.

When I ran locally for a version with a known issue (for a single build variant):

with known issue override: none   1.14s user 0.33s system 6% cpu 24.279 total
without known issue override:none   1.00s user 0.29s system 5% cpu 22.321 total

When running with no version override and no found versions at all (for the same single variant):

with known issue override: none   2.71s user 0.81s system 4% cpu 1:16.62 total
without known issue override: none   2.63s user 0.75s system 3% cpu 1:24.83 total

When running with some variant that has some known failures and no version override:
with known issue override: none   1.96s user 0.52s system 4% cpu 52.102 total -- version found
without known issue override: none   4.71s user 1.15s system 4% cpu 2:07.55 total -- timeout hit

I did have to add a condition to not check for a task that is not completed. That greatly improved performance (otherwise we would check if a task is a known issue even if it did not complete its run)

File: src/goodbase/services/evg_service.py [hunk 12]
```diff
@@ -92,7 +133,10 @@ def check_version(self, evg_version: Version, build_checks: List[BuildChecks]) -
         return success_check and (len(checks_with_failure_thresholds) == 0 or failure_check)
 
     def get_build_statuses_for_version(
-        self, evg_version: Version, build_checks: List[BuildChecks]
+        self,
+        evg_version: Version,
+        build_checks: List[BuildChecks],
+        allow_known_failures: bool = False,
```
Comment (zituo-jin) [inline src/goodbase/services/evg_service.py:139]: [nit] docstring

File: src/goodbase/services/search_service.py [hunk 13]
```diff
@@ -33,7 +33,11 @@ def __init__(
         self.options = options
 
     def find_revision(
-        self, evg_project: str, build_checks: List[BuildChecks], version_override: Optional[str]
+        self,
+        evg_project: str,
+        build_checks: List[BuildChecks],
+        version_override: Optional[str],
+        allow_known_failures: bool = False,
```
Comment (zituo-jin) [inline src/goodbase/services/search_service.py:40]: [nit] docstring

File: src/goodbase/services/search_service.py [hunk 14]
```diff
@@ -48,19 +52,26 @@ def find_revision(
...
+                )
 
         return stable_revision
 
     def _find_stable_revision(
-        self, evg_versions: Iterable[Version], build_checks: List[BuildChecks]
+        self,
+        evg_versions: Iterable[Version],
+        build_checks: List[BuildChecks],
+        allow_known_failures: bool = False,
```
Comment (zituo-jin) [inline src/goodbase/services/search_service.py:75]: [nit] docstring

File: src/goodbase/goodbase_cli.py [hunk 15]
```diff
@@ -138,6 +138,7 @@ def checkout_good_base(
         self,
         evg_project: str,
         build_checks: List[BuildChecks],
+        allow_known_failures: bool = False,
         version_override: Optional[str] = None,
```
Comment (zituo-jin) [inline src/goodbase/goodbase_cli.py:142]: [nit] docstring

**Review by zituo-jin — APPROVED** [review]
LGTM % a few nits
